text/0946-unified-imports.md

```diff
@@ -0,0 +1,151 @@
+---
+stage: accepted
+start-date: 2023-08-18T00:00:00.000Z 
+release-date: # In format YYYY-MM-DDT00:00:00.000Z
+release-versions:
+teams: 
+  - framework
+  - learning
+  - steering
+  - typescript
+prs:
+  accepted: # Fill this in with the URL for the Proposal RFC PR
+project-link:
+suite: 
+---
+
+<!--- 
+Directions for above: 
+
+stage: Leave as is
+start-date: Fill in with today's date, 2032-12-01T00:00:00.000Z
+release-date: Leave as is
+release-versions: Leave as is
+teams: Include only the [team(s)](README.md#relevant-teams) for which this RFC applies
+prs:
+  accepted: Fill this in with the URL for the Proposal RFC PR
+project-link: Leave as is
+suite: Leave as is
+-->
+
+# Unified imports for `<template>` 
+
+## Summary
+
+> One paragraph explanation of the feature.
+
+Ported from: https://github.com/emberjs/rfcs/issues/945
+
+
+I've been trying out [`ember-template-imports`](https://github.com/ember-template-imports/ember-template-imports) in my app, really liking this new way of defining components so far. Thanks to everyone who made it happen.
+
+One thing that's been bothering me slightly is the number of imports that some of my components need. I'm sure that better editor support will help, as will a possible future consolidation of core Ember imports, but I wish importing common things wasn't so scattered and verbose, eg:
+
+```js
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { service } from '@ember/service';
+import { on } from '@ember/modifier';
+import { array, hash, fn } from '@ember/helper';
+
+export default class extends Component {
+  @tracked isVisible = false;
+
+  <template>
+    ...
+  </template>
+}
+```
+
+I think there's an opportunity to simplify these imports. 
+
+In my own app, I've stared to re-export common built-ins:
+
+**`<my-app>/index.js`**
+```js
+export { default as Component } from '@glimmer/component';
+export { tracked } from '@glimmer/tracking';
+export { service } from '@ember/service';
+export { on } from '@ember/modifier';
+export { array, hash, fn } from '@ember/helper';
+```
+
+Which makes importing much simpler:
+
+**`<my-app>/components/modal.gjs`**
+```js
+import { Component, on, or, tracked } from 'vidu-web';
```

Comment:
I wouldn't want to do this, but I think I'd prefer something like:
```js
import Component from '@glimmer/component'; 
import { on, or } from '@glimmer/template'; 
import { tracked } from '@starbeam/ember';
```

where, 
`@glimmer/template` has:
- the types from `@glint/template`
  - `ModifierLike`
  - `HelperLike`
  - `ComponentLike`
  - `Invoke` (currently private, but is the only way ember-resources is glint compatible :sweat_smile:)
- eq
- notEq
- not
- get
- concat
- uniqueId
- page-title

and,
`@starbeam/ember` has:
- cell
- resource
- `@tracked`
- `@cached`
- ReactiveMap
- ReactiveSet
- ReactiveWeakMap
- ReactiveWeakSet


and these are built in, and no longer (and many already didn't) need importing:
- `on`
- `fn`
- `hash`
- `array`
- `hasBlock`
- `hasBlockParams`
- `helper` (for currying)
- `modifier` (for currying)
- `component` (for currying)
- `in-element`
- `log`
- `debugger`

These should be deprecated / removed, imo
- action
- input
- link-to
- mount
- mut
- textarea
- unbound

Reply:
Yea, I'm not a fan of the original proposal. I prefer clear import paths instead of importing everything from a single namespace to make it a _little_ bit easier. With proper tooling support there shouldn't be a difference anyway. 

I do think revisiting and streamlining the imports, as you suggest here, makes sense but that seems like a whole different discussion compared to what's suggested in the RFC? 😄

Reply:
I can see how importing from the originating source package can be considered clearer, but it brings with it downside of introducing multiple concepts like Glimmer, Starbeam etc. Every-time you need to import something, there's the small mental taxation of trying to remember which package it comes from.

I don't have a strong opinion on this, but from the point of view of someone new to Ember, I would think that importing common Ember constructs from an easy to remember place would simplify things greatly.

Reply:
We can incrementally get there, at the most basic, we should reorganize the tools anyway.

Once we have vite by default, and can _ensure_ that tree-shaking works, we could provide an `import { on, fn, eq, tracked, etc } from '@ember/meta';`

but I don't think it makes sense to strive for something like that unless we can make sure that unused functions are not shipped in folks apps.

Reply:
> but it brings with it downside of introducing multiple concepts like Glimmer, Starbeam etc

fwiw, all other frameworks have this split, so _at the very least_, we meet the status quo.

Given the above suggestion, we have only 3 concepts:
- a component (a feature / refactoring boundary for a collection of "things")
- the reactive primitives and utilities (starbeam)
- the template ulities (glimmer)